core/src/Dscp/Util/Timing.hs

```diff
@@ -0,0 +1,43 @@
+module Dscp.Util.Timing
+       ( countingTime
+       , periodically
+       ) where
+
+import Control.Concurrent (threadDelay)
+import Data.Time.Clock.POSIX (getPOSIXTime)
+import Fmt ((+|), (+||), (|+), (||+))
+import Loot.Log (MonadLogging, logError)
+import Time (KnownRat, Second, Time, toNum)
+import UnliftIO (MonadUnliftIO)
+
+import Dscp.Util.TimeLimit (logWarningWaitOnce)
+
+-- | Evaluate how much time did action take.
+countingTime :: MonadIO m => m a -> m (Double, a)
+countingTime m = do
+    t <- liftIO $ getPOSIXTime
+    a <- m
+    t' <- liftIO $ getPOSIXTime
+    let diff :: Double
+        diff = fromRational . toRational $ t' - t
+    return (diff, a)
+
+-- | Execute an action periodically till the end of node's days.
+periodically
+    :: (MonadUnliftIO m, MonadCatch m, MonadLogging m, KnownRat unit)
+    => Text -> Time unit -> m () -> m a
+periodically desc period action = loop
+  where
+    periodSec = toNum @Second period
+    loop = do
+        start <- liftIO getPOSIXTime
+        safeAction
+        end <- liftIO getPOSIXTime
+        let remaining = periodSec - (end - start)
+        liftIO $ threadDelay (round $ remaining * 1000000)
+        loop
+    safeAction =
+        logWarningWaitOnce period desc $
+        action `catchAny` reportFailure
```

Comment:
I don't think periodic action should stop being repeated on first exception. Use `foreverAlive` here instead, or better - the version of `foreverAlive` which makes sure that recovery delay after catching an exception is such that next action attempt will be performed in the beginning of next period.

Reply:
Mm, actually I tried to merge `periodically` and `foreverAlive`, and got something unpleasant. These functions are too different: exceptions are handles in distinct ways, and one waits for a given time disregard possible exceptions while another executes the action at max rate.

And current `periodically` function tolerates exceptions.